File: taurus-protect-sdk-python/taurus_protect/models/address.py
"""Address models for Taurus-PROTECT SDK."""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field

from taurus_protect.models.balance import Balance


class AddressAttribute(BaseModel):
    """Custom attribute on an address."""

    id: str = Field(description="Attribute identifier")
    key: str = Field(description="Attribute name")
    value: str = Field(description="Attribute value")

    model_config = {"frozen": True}


class Address(BaseModel):
    """
    Blockchain address within a wallet.

    Attributes:
        id: Unique address identifier.
        wallet_id: ID of the containing wallet.
        address: The blockchain address string.
        alternate_address: Alternate address format if available.
        label: Human-readable label.
        comment: Optional description.
        currency: Currency symbol.
        customer_id: Optional customer identifier.
        external_address_id: Optional external identifier.
        address_path: HD derivation path.
        address_index: Index used for address generation.
        nonce: Current address nonce.
        status: Creation status (created, creating, signed, observed, confirmed).
        balance: Current address balance.
        signature: Cryptographic signature for integrity verification.
        disabled: Whether the address is disabled.
        can_use_all_funds: Whether all funds can be used.
        created_at: When the address was created.
        updated_at: When the address was last updated.
        attributes: Custom key-value attributes.
        linked_whitelisted_address_ids: IDs of linked whitelisted addresses.
    """

    id: str = Field(description="Unique address identifier")
    wallet_id: str = Field(description="Containing wallet ID")
    address: str = Field(description="Blockchain address string")
    alternate_address: Optional[str] = Field(default=None, description="Alternate format")
    label: Optional[str] = Field(default=None, description="Human-readable label")
    comment: Optional[str] = Field(default=None, description="Optional description")
    currency: str = Field(default="", description="Currency symbol")
    customer_id: Optional[str] = Field(default=None, description="Customer identifier")
    external_address_id: Optional[str] = Field(default=None, description="External ID")
    address_path: Optional[str] = Field(default=None, description="HD derivation path")
    address_index: Optional[str] = Field(default=None, description="Address index")
    nonce: Optional[str] = Field(default=None, description="Current nonce")
    status: Optional[str] = Field(default=None, description="Creation status")
    balance: Optional[Balance] = Field(default=None, description="Current balance")
    signature: Optional[str] = Field(default=None, description="Integrity signature")
    disabled: bool = Field(default=False, description="Whether disabled")
    can_use_all_funds: bool = Field(default=False, description="Can use all funds")
    created_at: Optional[datetime] = Field(default=None, description="Creation timestamp")
    updated_at: Optional[datetime] = Field(default=None, description="Update timestamp")
    attributes: List[AddressAttribute] = Field(
        default_factory=list, description="Custom attributes"
    )
    linked_whitelisted_address_ids: List[str] = Field(
        default_factory=list, description="Linked whitelisted address IDs"
    )

    model_config = {"frozen": True}


class CreateAddressRequest(BaseModel):
    """Request to create a new address."""

    wallet_id: str = Field(description="Wallet ID (required)")
    label: str = Field(description="Address label (required)")
    comment: Optional[str] = Field(default=None, description="Optional description")
    customer_id: Optional[str] = Field(default=None, description="Customer identifier")
    external_address_id: Optional[str] = Field(default=None, description="External ID")
    address_type: Optional[str] = Field(
        default=None, description="Address type (e.g., p2pkh, p2wpkh)"
    )
    non_hardened_derivation: bool = Field(default=False, description="Use non-hardened derivation")


class ListAddressesOptions(BaseModel):
    """Options for listing addresses."""

    wallet_id: Optional[str] = Field(default=None, description="Filter by wallet ID")
    limit: int = Field(default=50, ge=1, le=1000, description="Maximum items to return")
    offset: int = Field(default=0, ge=0, description="Number of items to skip")
    query: Optional[str] = Field(default=None, description="Search query")
    exclude_disabled: bool = Field(default=False, description="Exclude disabled addresses")

File: taurus-protect-sdk-typescript/src/models/address.ts
/**
 * Address models for Taurus-PROTECT SDK.
 */

import type { OffsetPageOptions } from "./pagination";

/**
 * Address status enum.
 */
export enum AddressStatus {
  UNKNOWN = "UNKNOWN",
  CREATED = "CREATED",
  CREATING = "CREATING",
  SIGNED = "SIGNED",
  OBSERVED = "OBSERVED",
  CONFIRMED = "CONFIRMED",
  ACTIVE = "ACTIVE",
  PENDING = "PENDING",
  DISABLED = "DISABLED",
}

/**
 * Custom attribute on an address.
 */
export interface AddressAttribute {
  /** Attribute identifier */
  readonly id: string;
  /** Attribute name */
  readonly key: string;
  /** Attribute value */
  readonly value: string;
}

/** Balance information for an address. */
export interface Balance {
  readonly totalConfirmed: string;
  readonly totalUnconfirmed: string;
  readonly availableConfirmed: string;
  readonly availableUnconfirmed: string;
  readonly reservedConfirmed: string;
  readonly reservedUnconfirmed: string;
}

/**
 * Represents a blockchain address in Taurus-PROTECT.
 *
 * Addresses are created within wallets and are used to send and receive
 * cryptocurrency. Each address has an HSM-generated signature that can
 * be verified to ensure integrity.
 */
export interface Address {
  /** Unique address identifier */
  readonly id: string;
  /** Parent wallet ID */
  readonly walletId: string;
  /** The blockchain address string */
  readonly address: string;
  /** Alternate address format if available */
  readonly alternateAddress: string | undefined;
  /** Human-readable label */
  readonly label: string | undefined;
  /** Optional description */
  readonly comment: string | undefined;
  /** Currency symbol (e.g., 'ETH', 'BTC') */
  readonly currency: string;
  /** Optional customer identifier */
  readonly customerId: string | undefined;
  /** Optional external identifier */
  readonly externalAddressId: string | undefined;
  /** HD derivation path */
  readonly addressPath: string | undefined;
  /** Index used for address generation */
  readonly addressIndex: string | undefined;
  /** Current address nonce */
  readonly nonce: string | undefined;
  /** Creation status (created, creating, signed, observed, confirmed) */
  readonly status: string | undefined;
  /** HSM signature for integrity verification */
  readonly signature: string | undefined;
  /** Whether the address is disabled */
  readonly disabled: boolean;
  /** Whether all funds can be used */
  readonly canUseAllFunds: boolean;
  /** Creation timestamp */
  readonly createdAt: Date | undefined;
  /** Last update timestamp */
  readonly updatedAt: Date | undefined;
  /** Custom key-value attributes */
  readonly attributes: AddressAttribute[];
  /** IDs of linked whitelisted addresses */
  readonly linkedWhitelistedAddressIds: string[];
  /** Balance information */
  readonly balance?: Balance;
}

/**
 * Request to create a new address.
 */
export interface CreateAddressRequest {
  /** Wallet ID (required) */
  readonly walletId: string;
  /** Address label (required) */
  readonly label: string;
  /** Optional description */
  readonly comment?: string;
  /** Optional customer identifier */
  readonly customerId?: string;
  /** Optional external identifier */
  readonly externalAddressId?: string;
  /** Address type (e.g., p2pkh, p2wpkh) */
  readonly addressType?: string;
  /** Use non-hardened derivation */
  readonly nonHardenedDerivation?: boolean;
}

/**
 * Options for listing addresses. An offset list: `limit` 1-100 (default 20) and
 * `offset` (a previous page's `pagination.nextOffset`).
 */
export interface ListAddressesOptions extends OffsetPageOptions {
  /** Filter by wallet ID */
  readonly walletId?: string;
  /** Search query */
  readonly query?: string;
  /** Filter by blockchain */
  readonly blockchain?: string;
  /** Filter by network */
  readonly network?: string;
  /** Filter by Taurus address IDs (at most 50) */
  readonly addressIds?: string[];
  /** Filter by blockchain addresses (at most 100) */
  readonly addresses?: string[];
  /** Keep addresses carrying any of these tags */
  readonly tagIds?: string[];
  /** Keep only addresses with a positive balance */
  readonly onlyPositiveBalance?: boolean;
  /** Keep addresses whose balance is above this amount */
  readonly balanceAbove?: string;
  /** Keep addresses whose balance is below this amount */
  readonly balanceBelow?: string;
  /** Sort field */
  readonly sortBy?: string;
  /** Sort order ("ASC" or "DESC") */
  readonly sortOrder?: string;
  /**
   * Leave disabled addresses out (sent as `includeDisabledAddresses=exclude`). Unset or
   * false keeps the server default, which includes them.
   */
  readonly excludeDisabled?: boolean;
}
